### PR TITLE
Add discriminator to vits

### PR DESCRIPTION
**Pull request by ylacombe**
# What does this PR do?

Vits is a TTS model that have been supported in HF for a few months now. MMS-tts, Meta's models on a thousand languages, also uses VITS architecture.

Vits is a peculiar model, a GAN, a VAE a flow-based, all at the same time. To train it, one needs a discriminator, which this PR adds to Transformers.

**This PR aims to add VitsDiscriminator and a dedicated model for training VITS. PR #27244 will then add a training script example.**

Main changes:

- adding VitsDiscriminator
- adding VitsModelForPretraining
- allows passing multiple speaker_id, instead of a single one. Works if nb_speaker_id = batch_size
- weight_norm functions - to keep aligned to the original code
- Resizing the speaker embeddings feature
- I haven't touched VitsModel except for the handling of multiple speaker id in the forward method.

cc @amyeroberts and @sanchit-gandhi ! This is a separate PR as requested!

### CODE REVIEW

File: src/transformers/models/vits/modeling_vits.py [hunk 1]
```diff
@@ -430,6 +505,131 @@ def forward(self, inputs, padding_mask, global_conditioning=None):
...
+
+        cond_layer = torch.nn.Conv1d(speaker_embedding_size, 2 * hidden_size * num_layers, 1)
+        self.wavenet.cond_layer = nn.utils.weight_norm(cond_layer, name="weight")
+        nn.init.kaiming_normal_(self.wavenet.cond_layer.weight)
+        if self.wavenet.cond_layer.bias is not None:
+            k = math.sqrt(
+                self.wavenet.cond_layer.groups
+                / (self.wavenet.cond_layer.in_channels * self.wavenet.cond_layer.kernel_size[0])
+            )
+            nn.init.uniform_(self.wavenet.cond_layer.bias, a=-k, b=k)
```
Comment (amyeroberts) [inline src/transformers/models/vits/modeling_vits.py:527]: Here I think we should be using self._init_weights(cond_layer). Is the logic different from the weight initialization in the pretrained model?

File: src/transformers/models/vits/modeling_vits.py [hunk 2]
```diff
@@ -527,6 +727,14 @@ def __init__(self, config: VitsConfig):
         if config.speaker_embedding_size != 0:
             self.cond = nn.Conv1d(config.speaker_embedding_size, config.upsample_initial_channel, 1)
 
+    def resize_speaker_embedding(self, speaker_embedding_size):
+        self.config.speaker_embedding_size = speaker_embedding_size
+        self.cond = nn.Conv1d(speaker_embedding_size, self.config.upsample_initial_channel, 1)
+        nn.init.kaiming_normal_(self.cond.weight)
+        if self.cond.bias is not None:
+            k = math.sqrt(self.cond.groups / (self.cond.in_channels * self.cond.kernel_size[0]))
+            nn.init.uniform_(self.cond.bias, a=-k, b=k)
```
Comment (amyeroberts) [inline src/transformers/models/vits/modeling_vits.py:736]: Same comment here about the initialization

File: src/transformers/models/vits/modeling_vits.py [hunk 3]
```diff
@@ -619,6 +837,30 @@ def forward(self, inputs, padding_mask, global_conditioning=None, reverse=False)
...
+
+            cond_layer = torch.nn.Conv1d(speaker_embedding_size, 2 * hidden_size * num_layers, 1)
+            flow.wavenet.cond_layer = nn.utils.weight_norm(cond_layer, name="weight")
+            nn.init.kaiming_normal_(flow.wavenet.cond_layer.weight)
+            if flow.wavenet.cond_layer.bias is not None:
+                k = math.sqrt(
+                    flow.wavenet.cond_layer.groups
+                    / (flow.wavenet.cond_layer.in_channels * flow.wavenet.cond_layer.kernel_size[0])
+                )
+                nn.init.uniform_(flow.wavenet.cond_layer.bias, a=-k, b=k)
```
Comment (amyeroberts) [inline src/transformers/models/vits/modeling_vits.py:862]: And here

File: src/transformers/models/vits/modeling_vits.py [hunk 4]
```diff
@@ -760,6 +1002,15 @@ def __init__(self, config):
...
             self.post_flows.append(VitsConvFlow(config))
 
+        self.filter_channels = filter_channels
+
+    def resize_speaker_embeddings(self, speaker_embedding_size):
+        self.cond = nn.Conv1d(speaker_embedding_size, self.filter_channels, 1)
+        nn.init.kaiming_normal_(self.cond.weight)
+        if self.cond.bias is not None:
+            k = math.sqrt(self.cond.groups / (self.cond.in_channels * self.cond.kernel_size[0]))
+            nn.init.uniform_(self.cond.bias, a=-k, b=k)
```
Comment (amyeroberts) [inline src/transformers/models/vits/modeling_vits.py:1012]: And here

File: src/transformers/models/vits/modeling_vits.py [hunk 5]
```diff
@@ -1487,3 +1766,420 @@ def forward(
...
+
+            attn_mask = torch.unsqueeze(input_padding_mask, 2) * torch.unsqueeze(labels_padding_mask, -1)
+
+            attn = monotic_alignment_function(neg_cent, attn_mask.squeeze(1)).unsqueeze(1).detach()
+
+        durations = attn.sum(2)
+
+        if self.config.use_stochastic_duration_prediction:
+            log_duration = self.duration_predictor(
+                hidden_states, input_padding_mask, speaker_embeddings, durations=durations, reverse=False
```
Comment (amyeroberts) [inline src/transformers/models/vits/modeling_vits.py:2126]: How come self.noise_scale_duration isn't passed in here?

File: src/transformers/models/vits/modeling_vits.py [hunk 6]
```diff
@@ -1487,3 +1766,420 @@ def forward(
...
+                    f"You passed {len(speaker_id)} `speaker_id` but you should either pass one speaker id or `batch_size` `speaker_id`."
+                )
+
+            speaker_embeddings = self.embed_speaker(speaker_id).unsqueeze(-1)
+        else:
+            speaker_embeddings = None
+
+        # if inference, return inference forward of VitsModel
+        if labels is None:
+            return self._inference_forward(
```
Comment (amyeroberts) [inline src/transformers/models/vits/modeling_vits.py:2065]: Is it necessary to have a whole separate method here? It seems there's a lot of overlap in the logic. It also breaks the common `forward` pattern in other models and makes the logic overall harder to maintain as it has to be updated in two places

**Review by amyeroberts — COMMENTED** [review]
Nice - thanks for adding! 

Some nits and structural questions but overall looks good.

File: src/transformers/models/vits/feature_extraction_vits.py [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+
+        # convert into correct format for padding
+        padded_inputs = self.pad(
+            batched_speech,
+            padding=padding,
+            max_length=max_length,
+            truncation=truncation,
+            pad_to_multiple_of=pad_to_multiple_of,
+            return_attention_mask=return_attention_mask or do_normalize,
+            return_tensors="pt",
```
[the remark this answers is not in the document]
Comment (ylacombe) [reply src/transformers/models/vits/feature_extraction_vits.py:236]: Yes it is, I'll add the `require_backends`, thanks

[the remark this answers is not in the document]
Comment (ylacombe) [reply src/transformers/models/vits/feature_extraction_vits.py:236]: Well I can't find `require_backends` ! Do you mean `require_torch` ? I though it was specific to testing

File: src/transformers/models/vits/modeling_vits.py (hunk 5 above)
Comment (ylacombe) [reply src/transformers/models/vits/modeling_vits.py:2126]: It's because it's in training mode (i.e `reverse=False`). In that case the duration predictor doesn't use `noise_scale`

File: src/transformers/models/vits/modeling_vits.py (hunk 6 above)
Comment (ylacombe) [reply src/transformers/models/vits/modeling_vits.py:2065]: Well, for a few reasons:
1. the original repo has a separate method, I simply replicated the logic
2. there are IMO a lot of difference
3. it was easier to debug !

I could indeed fuse the two methods (to train and to infer) but IMO, it would be even more difficult to maintain - here we only have to replicate part of the `VitsModel.forward` on `_inference_forward`. If I fused the two methods, it would be more difficult for sure.

I'll listen to your suggestions of course! What do you think would be the best option?
1. Fuse training and inference logic
2. Only have the training logic in `VitsModelForPretraining` and leave the inference code for `VitsModel`
3. Leave it as it is

-------

I'll also put the inference and train diagram here just in case for reference:

**TRAINING**

image

**INFERENCE**

image

File: src/transformers/models/vits/modeling_vits.py [hunk 8]
```diff
@@ -123,6 +127,64 @@ def fused_add_tanh_sigmoid_multiply(input_a, input_b, num_channels):
     return acts
 
 
+def slice_segments(hidden_states, ids_str, segment_size=4):
+    ret = torch.zeros_like(hidden_states[:, :, :segment_size])
+    for i in range(hidden_states.size(0)):
```
[the remark this answers is not in the document]
Comment (ylacombe) [reply src/transformers/models/vits/modeling_vits.py:132]: No, you're right! A gather should do the work!

File: src/transformers/models/vits/modeling_vits.py (hunk 6 above)
Comment (amyeroberts) [reply src/transformers/models/vits/modeling_vits.py:2065]: OK, I see, thanks for sharing the diagram! 

In general, we shouldn't just be following the design patterns of other repos. Part of the advantage of transformers is that it provides a standardised interface and code structure - so sometimes we have to re-work quite a bit to keep this. 

That being said, if there's large differences, we don't want to make the forward pass too hard to understand. I'm happy to leave as-is 👍

File: src/transformers/models/vits/modeling_vits.py (hunk 1 above)
Comment (ylacombe) [reply src/transformers/models/vits/modeling_vits.py:576]: You're right, good catch!
